lib/balrog/middleware.rb

```diff
@@ -59,27 +82,77 @@ def handle_login(env)
       EOF
     end
 
-    if @password_hash == submitted_password
-      session_data = { value: 'authenticated' }
-      add_expiry_date!(session_data)
-      env['rack.session'][:balrog] = session_data
+    if password_hash == submitted_password
+      authenticate_user(env)
     end
 
     referer = env["HTTP_REFERER"] || '/'
 
     [302, {"Location" => referer}, [""]]
   end
 
+  def handle_omniauthentication(env)
+    if env['omniauth.auth']['info']['email']
+      user_email = env['omniauth.auth']['info']['email']
+      email_domain = user_email.split("@").last
+    end
+    
+    unless email_domain
```

Comment:
Add some comments for each of these conditional guards, e.g.

`# Double check the email address has a domain before we check the whitelist`